### PR TITLE
[Form] add a convenience method to get the parent form in Twig templates

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28686
| License       | MIT
| Doc PR        | symfony/symfony-docs#10999

### CODE REVIEW

**Comment by ro0NL** [comment]
it's a BC break no? also ref https://github.com/symfony/symfony/pull/19492#issuecomment-236418002

**Comment by xabbuh** [comment]
Why would that be a BC break?

**Comment by xabbuh** [comment]
Taking into account https://github.com/symfony/symfony/pull/19492#issuecomment-236418002 we could of course think about adding a Twig function instead if we do not want to add this method to the `FormView` class.

**Comment by ro0NL** [comment]
the order changes right? before a child named parent might be returned first, whereas now it's always the parent form

**Comment by ro0NL** [comment]
alternatively, getting rid of arrayaccess in formview would also do it towards consistent accessors in php as well as twig, but that's a bigger step :)

**Comment by xabbuh** [comment]
If you refer to `form.parent`, this still returns the child form as Twig considers `ArrayAccess` methods before looking at actually implemented methods. And `form.getParent()` just is not usable without this change.

> alternatively, getting rid of arrayaccess in formview would also do it towards consistent accessors in php as well as twig, but that's a bigger step :)

I don't think that's a good idea DX wise. Being able to use `form.child_name` is quite convenient.

**Comment by yceruto** [comment]
@xabbuh could we deprecating `twig_is_root_form()` test func then? I guess it wouldn't useful anymore, replacing it by `form.getParent() is null`.

---

In the other hand, I think the solution should be done in Twig context, because having a getter for a public property seems strange.

What about to add a new Twig function for access to the `FormView` public properties? Something like this:
```php
public function getPropertyValue($object, string $propertyName)
{
        return $object->$propertyName;
}
```
for cases where the public property matches the name of the child form (`ArrayAccess` ambiguity in Twig context):
```twig
{# $formView->parent #}
{% if property_value(form, 'parent') is null %}

{# $formView->vars #}
{% if property_value(form, 'vars').attr.class is defined %}
```
even, other `ArrayAccess` objects (in userland) with a composition similar to `FormView` can benefit from it, too.

**Comment by xabbuh** [comment]
I have rewritten this solution to add a new `parent_form()` Twig function instead. I would not deprecate the filter for now. Both the filter and the new function serve their own use cases.

File: src/Symfony/Bridge/Twig/Extension/FormExtension.php [hunk 1]
```diff
@@ -115,3 +116,11 @@ function twig_is_root_form(FormView $formView)
 {
     return null === $formView->parent;
 }
+
+/**
+ * @internal
+ */
+function twig_get_root_form(FormView $formView): ?FormView
```
Comment (yceruto) [inline src/Symfony/Bridge/Twig/Extension/FormExtension.php]: Did you mean `twig_get_parent_form`?

Comment (xabbuh) [reply src/Symfony/Bridge/Twig/Extension/FormExtension.php]: :+1: good catch, fixed

**Review by ro0NL — APPROVED** [review]

**Review by yceruto — APPROVED** [review]

**Comment by javiereguiluz** [comment]
@xabbuh please, create an issue in Symfony Docs to not forget about this. Also, please add some usage examples and if this improves an existing feature add a before/after example. Finally, please check if this fixes something that the current docs say it's not possible to do or buggy. Thanks!

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by xabbuh** [comment]
@javiereguiluz it took me some time, but here we go now: symfony/symfony-docs#10999

**Pull request merged**